server/bleep/src/cache.rs

```diff
@@ -221,93 +197,170 @@ impl<'a> ChunkCache<'a> {
-                .await
-        }))
-        .await
-        .into_iter()
-        .collect::<Result<Vec<_>, _>>()?;
+    pub async fn commit(self, qdrant: &QdrantClient) -> anyhow::Result<(usize, usize, usize)> {
+        let mut update_size = 0;
+        let mut updates = vec![];
+        let mut tx = self.sql.begin().await?;
+
+        //
+        // Update points
+        //
+        let mut next = self.update.first_occupied_entry();
+        while let Some(entry) = next {
+            let (branches_list, branches) = entry.key();
+            let points = entry.get();
+            update_size += points.len();
+
+            for p in entry.get() {
+                sqlx::query! {
+                    "UPDATE chunk_cache SET branches = ? \
+                     WHERE chunk_hash = ?",
+                     branches,
+                     p
+                }
+                .execute(&mut *tx)
+                .await?;
+            }
 
+            let id = points
+                .iter()
+                .cloned()
+                .map(PointId::from)
+                .collect::<Vec<_>>()
+                .into();
+            let payload = qdrant_client::client::Payload::new_from_hashmap(
+                [("branches".to_string(), branches_list.to_owned().into())].into(),
+            );
+
+            updates.push(async move {
+                qdrant
+                    .set_payload_blocking(semantic::COLLECTION_NAME, &id, payload, None)
+                    .await
+            });
+            next = entry.next();
+        }
+        futures::future::join_all(updates.into_iter())
+            .await
+            .into_iter()
+            .collect::<Result<Vec<_>, _>>()?;
+
+        //
+        // Delete points
+        //
         let mut to_delete = vec![];
         self.cache
             .scan_async(|id, p| {
                 if !p.fresh {
-                    to_delete.push(PointId::from(id.to_owned()))
+                    to_delete.push(id.to_owned());
                 }
             })
             .await;
 
         let delete_size = to_delete.len();
+        for p in to_delete.iter() {
+            sqlx::query! {
+                "DELETE FROM chunk_cache \
+                 WHERE chunk_hash = ? AND file_hash = ?",
+                p,
+                self.file_cache_key
+            }
+            .execute(&mut *tx)
+            .await?;
+        }
         if !to_delete.is_empty() {
-            self.qdrant
-                .delete_points(semantic::COLLECTION_NAME, &to_delete.into(), None)
+            qdrant
+                .delete_points(
+                    semantic::COLLECTION_NAME,
+                    &to_delete
+                        .into_iter()
+                        .map(PointId::from)
+                        .collect::<Vec<_>>()
+                        .into(),
+                    None,
+                )
                 .await?;
         }
 
+        //
+        // Upsert points
```

Comment:
Why is this an `upsert`? Aren't these just new points?

Reply:
They should be, but this operation has always been `upsert`, so I didn't change it. We don't really require strict semantics that ensure these be new, all we're saying is "we want these to exist like this".